secboot/secboot_tpm.go

```diff
@@ -69,3 +95,314 @@ func checkSecureBootEnabled() error {
+	for _, chain := range params.EFILoadChains {
+		policyParams.LoadSequences = append(policyParams.LoadSequences, buildLoadSequence(chain))
+	}
+	if err := sbAddEFISecureBootPolicyProfile(pcrProfile, &policyParams); err != nil {
+		return fmt.Errorf("cannot add EFI secure boot policy profile: %v", err)
+	}
+
+	// Add systemd EFI stub profile
+	if len(params.KernelCmdlines) != 0 {
+		systemdStubParams := sb.SystemdEFIStubProfileParams{
+			PCRAlgorithm:   tpm2.HashAlgorithmSHA256,
+			PCRIndex:       tpmPCR,
+			KernelCmdlines: params.KernelCmdlines,
+		}
+		if err := sbAddSystemdEFIStubProfile(pcrProfile, &systemdStubParams); err != nil {
+			return fmt.Errorf("cannot add systemd EFI stub profile: %v", err)
+		}
+	}
+
+	// Add snap model profile
+	if len(params.Models) != 0 {
+		snapModelParams := sb.SnapModelProfileParams{
+			PCRAlgorithm: tpm2.HashAlgorithmSHA256,
+			PCRIndex:     tpmPCR,
+			Models:       params.Models,
+		}
+		if err := sbAddSnapModelProfile(pcrProfile, &snapModelParams); err != nil {
+			return fmt.Errorf("cannot add snap model profile: %v", err)
+		}
+	}
+
+	// Provision the TPM as late as possible
+	if err := tpmProvision(tpm, params.TPMLockoutAuthFile); err != nil {
+		return err
+	}
+	// Seal key to the TPM
+	creationParams := sb.KeyCreationParams{
+		PCRProfile: pcrProfile,
+		PINHandle:  pinHandle,
+	}
+	if err := sbSealKeyToTPM(tpm, key[:], params.KeyFile, params.PolicyUpdateDataFile, &creationParams); err != nil {
+		return fmt.Errorf("cannot seal data: %v", err)
+	}
+
+	return nil
+
+}
+
+func tpmProvision(tpm *sb.TPMConnection, lockoutAuthFile string) error {
+	// Create and save the lockout authorization file
+	lockoutAuth := make([]byte, 16)
+	// crypto rand is protected against short reads
+	_, err := rand.Read(lockoutAuth)
+	if err != nil {
+		return fmt.Errorf("cannot create lockout authorization: %v", err)
+	}
+	if err := osutil.AtomicWriteFile(lockoutAuthFile, lockoutAuth, 0600, 0); err != nil {
+		return fmt.Errorf("cannot write the lockout authorization file: %v", err)
+	}
+
+	// TODO:UC20: ideally we should ask the firmware to clear the TPM and then reboot
+	//            if the device has previously been provisioned, see
+	//            https://godoc.org/github.com/snapcore/secboot#RequestTPMClearUsingPPI
+	if err := sbProvisionTPM(tpm, sb.ProvisionModeFull, lockoutAuth); err != nil {
+		return fmt.Errorf("cannot provision TPM: %v", err)
+	}
+	return nil
+}
+
+func buildLoadSequence(filePaths []string) *sb.EFIImageLoadEvent {
+	// The idea of EFIImageLoadEvent is to build a set of load paths for the current
+	// device configuration. So you could have something like this:
+	//
+	// shim -> recovery grub -> recovery kernel 1
+	//                      |-> recovery kernel 2
+	//                      |-> recovery kernel ...
+	//                      |-> normal grub -> run kernel good
+	//                                     |-> run kernel try
+	//
+	// Or it could look like this, which is the same thing:
```

Comment:
If the illustrations before and after are meant to explain something that is lost to me. If they are equivalent perhaps only one variant could be retained. Alternatively, please highlight the key difference.

Reply:
The EFI binary load chains are implemented in `snapcore/secboot` in a tree structure. The commentary clarifies the fact that this data can be supplied as a real hierarchical tree, like in the first example, or in a flattened form listing individual combinations as shown in the second example. We currently work with a single `shim -> grub1 -> grub2 -> kernel` EFI binary load chain so there's only one possible representation, but this may change in the future.